Add modules support (with order resolution, etc.) and 'import std' support for Meson (initial draft, tested in Clang)

germandiagogomez opened the pull request:
EDIT: updated top comment. I will keep it up to date with changes added. Please read again since there are changes since yesterday. LAST UPDATE: september, the 8th, 2025

-----------------------------------

### Changes: 8th september 2025

Test project is here: [Nesecidad-0.0.2.zip](https://github.com/user-attachments/files/22200566/Nesecidad-0.0.2.zip)

     - more intuitive file conventions for modules.
     - project attachment updated: now it uses module interface unit, 
       module partition units and a single module implementation unit.
     

Add module partitions and implementation modules.

This commit adds support for implementation partitions
and modifies file conventions for a modularized target.

The conventions are:

  - your primary interface unit for your module should always
    be called 'module.cppm'. It should export module 'target-name'.

  - your interface partitions can have any name supported by a module.
    For example: MyThings.cppm. The interface partition
    should declare 'export module target-name:MyThings'.

  - Your *importable* interface implementation units should end with
    'Impl.cppm'. For example 'MyThingsImpl.cppm'. This should
    'module target-name:MyThings' (without export).

  - Your *non-importable* implementation can have any name *with an
    extension .cpp*, not a .cppm, since implementation units are
    not importable. It is highly recommended, though, that if you
    have a single implementation file, it is called moduleImpl.cpp.

  You can see the project attached in Github draft PR:
  https://github.com/mesonbuild/meson/pull/14989 as an example that uses
  interface partitions, primary interface units, an implementation
  module (but not an implementation module partition, though it should
  work), multiple targets and import std.

  It can run a couple of tests when fully compiled and it resolves
  dependency order via scanner with Clang.

----- 

### Changes 7th september:

Test project is here: [Nesecidad-0.0.1.tar.gz](https://github.com/user-attachments/files/22196953/Nesecidad-0.0.1.tar.gz)

DRAFT: Add modules support via clang-scan-deps.

Tested on Clang 19 on MacOS.

BEWARE, WARNING: This is a drafty implementation
whose only goal was to see module resolution working
in a project. This is not representative of the
quality of the code I would be willing to submit,
but just a fast test.

That said: support for modules has been added consisting of:

  - a global (not per target) scan step via clang-scan-deps.
  - the scan + accumulate step are done at once via
    clang-scan-deps.
  - dependency order resolution for module imports should
    work.
  - the strategy to activate scanning right now is global,
    even subprojects are scanned, based on compile_commands.json
    generated. You need clang 17 and above though only
    Clang 19 has been tested.

Only primary interface units have been tested so far (no partitions, etc), 
but the dependency order resolution is already there. Probably some 
problems could arise in name conventions when adding these but the 
order resolution should work already.


The script depscanaccumulate.py *currently has the path to
clang-scan-deps hardcoded* and will not work unless you fix it
to point to your clang-scan-deps until I do some cleanups.

Please find attached a small project to work with
that I converted in the Github issue:
https://github.com/mesonbuild/meson/pull/14989.

The project is a NES emulator that *does almost nothing* since it was
an abandoned side-project, but for the sake of converting to modules
it has been invaluable.

There are many strategies to do the scanning:

    - per file.
    - per target.
    - globally (as it is currently done with this prototype)

Also, we could make a differentiation to split the files that
are modules from the normal source files, but I would not
recommend it. I think doing per-target scans all-or-nothing
per target should be enough.

Currently, the .pcm files are put at the top-level of the
build directory. This should go in some global per-project
cache to the best of my understanding.

There is a small trick now, which is not optimal, but it does
the job: the cpp_COMPILER rule is added -fmodule-output={the-pcm-file}.pcm
-fprebuilt-module-path={the-build-dir} in its ARGS and since not all files are
modules, this is surrounded by --start-no-unused-arguments and
--end-no-unused-arguments. This avoids the need to generate more
fancy or dynamic stuff, which I do not know even if it is possible
at all at the moment.

This method of compiling generates a .o file and a .pcm as a
side-effect. There is another way, potentially more parallel,
to build separately (via --precompile flag) a .pcm and a .o,
but again, this will do as a first step I guess...

I would expect that for big projects, unless clang-scan-deps is
really fast, there might be other strategies that are better for
incremental builds but something like this should do for now.

I would even expect that Meson could choose in a smart way
what strategy to adopt? (maybe)

Your targets are what you are already familiar with, this is
a transparent feature except for:

   1. the name of your build target should have the name of the module, something like 'my.module'
   2. you need to have a module interface file called 'my.module.cppm'.

Control to opt-out from module scanning, when the strategy is not
global scanning, should probably be added, but since I used
compile_commands.json + global scanning I did not get bothered, which
whould have deviated me from my goal.

   I updated and activated the rule for should_use_dyndeps() in the
   ninja backend that already existed in the code and generated
   the depscanaccumulate rule.
   Check the build.ninja file to see what it is generated.

   - I generate every target with a dependency on std
     (using cpp_import_std feature, which is active).

   - A single deps.dd file is generated every time a file is touched.

   - any file you touch should trigger a deps.dd scan.

 We could make a differentiation to split the files that
 are modules from the normal source files to do less agressive scanning,
 but I would not recommend it. I think doing per-target scans
 all-or-nothing per target should be enough since these tools should do
 bulk updates and not be invoked once per file anyway.

 Currently, the .pcm files are put at the top-level of the
 build directory. This should go in some global per-project
 cache to the best of my understanding.

 There is a small trick now, which is not optimal, but it does
 the job: the cpp_COMPILE rule is added -fmodule-output={the-pcm-file}.pcm
 -fprebuilt-module-path={the-build-dir} and since not all files are
 modules, this is surrounded by --start-no-unused-arguments and
 --end-no-unused-arguments. If the file is just a normal translation unit,
 a .pcm will not be generated even if it was specified.

 This avoids the need to generate more fancy or dynamic stuff, which I do
 not know even if it is possible at all at the moment.

Please find attached a small project to work with
that I converted. The project is a NES emulator that
*does almost nothing* since it was an abandoned side-project,
but for the sake of converting to modules has been invaluable.

-------------------------------------------------------------------------
Also add 'import std' support for Meson Clang/GCC.

This commit adds the ability for Meson targets for:

 - clang (tested in clang 19 homebrew MacOS)
 - gcc (tested in a Docker container with Linx GCC15)

The feature can be used by setting a new option 'cpp_import_std'
globally, which can be true or false.

By default the option is conservatively set to false.

If enabled, during the configuration phase, Meson checks that the
compiler supports building the std library as a module via compiler
version checks.

At build time, it adds a custom target to build the std bmi. This bmi is
wrapped in a dependency and included in build targets that are
executables, shared libraries and libraries if the target uses c++
language and it is not in a subproject. See build_target function.

A best effort has been done to align debug/release flags with the built
version of the std library, though this needs review.

Note that in order to use the feature, the minimum version is C++23,
since import std is a C++23 feature.

An example of how to use it can be found in test '1 import std'.

Smoke-tested in:

   - Clang 19 homebrew, in my Mac.
   - Gcc 15 in Docker container.
   - Tested with a multi-file project that uses subprojects
     by replacing all std library with 'import std'.

ROADMAP:

  - Add override to build targets to avoid using 'import std'?

jpakkane commented on test cases/cpp/1 import std/meson.build.clang:
```diff
@@ -0,0 +1,42 @@
+)
+
+cpp20_std_module_dep = declare_dependency(
+  compile_args: ['-fmodule-file=std=@0@'.format(cpp20_std_module.full_path())],
```
None of this should be in the build files. These need to be handled in the background. For comparison, precompiled headers work by just adding `cpp_pch: 'foo/bar.h'`, everything else is done behind the scenes. The "actual code" that does things is in `ninjabackend.py`.

I don't have a clear idea what the API should actually be, but maybe something along the lines of:

```meson
project('foo', 'cpp', 
    default_options: ['cpp_std=c++26', 'cpp_modules=std'])
executable('importstd', 'importstd.cpp')
```

germandiagogomez replied:
This is how it works (I made some changes based on feedback from the chat in the internals of how it works, but externally it goes like this):

```
project(
  'import std',
  'cpp',
  version : '0.1',
  meson_version : '>= 1.3.0',
  default_options : ['warning_level=3', 'cpp_std=c++23',
                     'cpp_import_std=true'],
)


exe = executable(
  'import std',
  'import_std.cpp'
)
```

germandiagogomez committed:
DRAFT: Add modules support via clang-scan-deps.

Tested on Clang 19 on MacOS.

BEWARE, WARNING: This is a drafty implementation
whose only goal was to see module resolution working
in a projet. This is not representative of the
quality of the code I would be willing to submit,
but just a fast test.

That said: support for modules has been added consisting of:

  - a global (not per target) scan step via clang-scan-deps.
  - the scan + accumulate step are done at once via
    clang-scan-deps.
  - dependency order resolution for module imports should
    work.
  - the strategy to activate scanning right now is global,
    even subprojects are scanned, based on compile_commands.json
    generated. You need clang 17 and above though only
    Clang 19 has been tested.

The script depscanaccumulate.py *currently has the path to
clang-scan-deps hardcoded* and will not work unless you fix it
to point to your clang-scan-deps until I do some cleanups.

Please find attached a small project to work with
that I converted in the Github issue:
https://github.com/mesonbuild/meson/pull/14989.

The project is a NES emulator that *does almost nothing* since it was
an abandoned side-project, but for the sake of converting to modules
it has been invaluable.

There are many strategies to do the scanning:

    - per file.
    - per target.
    - globally (as it is currently done with this prototype)

Also, we could make a differentiation to split the files that
are modules from the normal source files, but I would not
recommend it. I think doing per-target scans all-or-nothing
per target should be enough.

Currently, the .pcm files are put at the top-level of the
build directory. This should go in some global per-project
cache to the best of my understanding.

There is a small trick now, which is not optimal, but it does
the job: the cpp_COMPILE rule is added -fmodule-output={the-pcm-file}.pcm
-fprebuilt-module-path={the-build-dir} and since not all files are
modules, this is surrounded by --start-no-unused-arguments and
--end-no-unused-arguments. This avoids the need to generate more
fancy or dynamic stuff, which I do not know even if it is possible
at all at the moment.

This method of compiling generates a .o file and a .pcm as a
side-effect. There is another way, potentially more parallel,
to build separately (via --precompile flag) a .pcm and a .o,
but again, this will do as a first step I guess...

I would expect that for big projects, unless clang-scan-deps is
really fast, there might be other strategies that are better for
incremental builds but something like this should do for now.

I would even expect that Meson could choose in a smart way
what strategy to adopt? (maybe)

Your targets are what you are already familiar with, this is
a transparent feature except for:

   1. the name of your build target should have the name of the module.
   2. you need to have a module interface file called target.name.cppm.

Control to opt-out from module scanning, when the strategy is not
global scanning, should probably be added, but since I used
compile_commands.json + global scanning I did not get bothered, which
whould have deviated me from my goal.

   I updated and activated the rule for should_use_dyndeps() in the
   ninja backend that already existed in the code and generated
   the depscanaccumulate rule.
   Check the build.ninja file to see what it is generated.

   - I generate every target with a dependency on std
     (using cpp_import_std feature, which is active).

   - A single deps.dd file is generated every time a file is touched.

   - any file you touch should trigger a deps.dd scan.

 We could make a differentiation to split the files that
 are modules from the normal source files to do less agressive scanning,
 but I would not recommend it. I think doing per-target scans
 all-or-nothing per target should be enough since these tools should do
 bulk updates and not be invoked once per file anyway.

 Currently, the .pcm files are put at the top-level of the
 build directory. This should go in some global per-project
 cache to the best of my understanding.

 There is a small trick now, which is not optimal, but it does
 the job: the cpp_COMPILE rule is added -fmodule-output={the-pcm-file}.pcm
 -fprebuilt-module-path={the-build-dir} and since not all files are
 modules, this is surrounded by --start-no-unused-arguments and
 --end-no-unused-arguments. If the file is just a normal translation unit,
 a .pcm will not be generated even if it was specified.

 This avoids the need to generate more fancy or dynamic stuff, which I do
 not know even if it is possible at all at the moment.

Please find attached a small project to work with
that I converted. The project is a NES emulator that
*does almost nothing* since it was an abandoned side-project,
but for the sake of converting to modules has been invaluable.

-------------------------------------------------------------------------
Also add 'import std' support for Meson Clang/GCC.

This commit adds the ability for Meson targets for:

 - clang (tested in clang 19 homebrew MacOS)
 - gcc (tested in a Docker container with Linx GCC15)

The feature can be used by setting a new option 'cpp_import_std'
globally, which can be true or false.

By default the option is conservatively set to false.

If enabled, during the configuration phase, Meson checks that the
compiler supports building the std library as a module via compiler
version checks.

At build time, it adds a custom target to build the std bmi. This bmi is
wrapped in a dependency and included in build targets that are
executables, shared libraries and libraries if the target uses c++
language and it is not in a subproject. See build_target function.

A best effort has been done to align debug/release flags with the built
version of the std library, though this needs review.

Note that in order to use the feature, the minimum version is C++23,
since import std is a C++23 feature.

An example of how to use it can be found in test '1 import std'.

Smoke-tested in:

   - Clang 19 homebrew, in my Mac.
   - Gcc 15 in Docker container.
   - Tested with a multi-file project that uses subprojects
     by replacing all std library with 'import std'.

ROADMAP:

  - Add override to build targets to avoid using 'import std'?

talisein commented:
I'm just offering some input as a user, but adding a project-wide option doesn't seem like it has any path to being usable for not-std modules in the future. I would imagine something like this would be more extensible for the future:

```
module_std = declare_cpp_module('std', ...)
lib_foo = static_library('foo', cpp_args: ['-DFOO'], cpp_modules: [module_std])
bar_exe = executable('bar', cpp_args: ['-DBAR'], cpp_modules: [module_std])
```

Where for a first step, declare_cpp_module() would just have some automagic for std, and later you can add arguments for sources for other modules.

And what needs to happen is the bmi is built once for lib_foo with lib_foo's cpp_args, and again for bar_exe with bar_exe's cpp_args. Any implementation that doesn't tie a bmi to the cpp_args is broken, isn't it?

jpakkane commented:
> adding a project-wide option doesn't seem like it has any path to being usable for not-std modules in the future

It would not be used for those. They would have some different mechanism, because they are a completely different thing (because they affect things like compilation step build orders etc).

Also note that those options can be set per-target. This would be equivalent to defining it at the top level:

```meson
project('foo', 'cpp', 
    default_options: ['cpp_std=c++26'])
executable('importstd', 'importstd.cpp',
    override_options: 'cpp_modules=std')
```

germandiagogomez commented:
> I'm just offering some input as a user, but adding a project-wide option doesn't seem like it has any path to being usable for not-std modules in the future. I would imagine something like this would be more extensible for the future:
> 
> ```
> module_std = declare_cpp_module('std', ...)
> lib_foo = static_library('foo', cpp_args: ['-DFOO'], cpp_modules: [module_std])
> bar_exe = executable('bar', cpp_args: ['-DBAR'], cpp_modules: [module_std])
> ```
> 
> Where for a first step, declare_cpp_module() would just have some automagic for std, and later you can add arguments for sources for other modules.
> 
> And what needs to happen is the bmi is built once for lib_foo with lib_foo's cpp_args, and again for bar_exe with bar_exe's cpp_args. Any implementation that doesn't tie a bmi to the cpp_args is broken, isn't it?

Currently you can do this:

```
project('foo', 'cpp', 
    default_options: ['cpp_std=c++26',  'cpp_import_std=true'])

executable('importstd', 'importstd.cpp')
```

which is global. But I plan to add per-target overrides, as @jpakkane mentioned:

project('foo', 'cpp', 
    default_options: ['cpp_std=c++26',  'cpp_import_std=true'])

executable('notimportstd', 'notimportstd.cpp', cpp_import_std: false)
```



By the way, looking at Clang documentation there is something I am not sure if I am doing wrong (because it works): it seems that every BMI file generates a .o file. However, right now I am just adding a target that generates a BMI for both gcc and clang and I compile by adding that custom target (with the flags for the build itself when compiling BMI).

But the library that is linked is the stock included libc++ to the best of my knowledge. 

So I am not sure if there is a potential mismatch between how I compiled the BMI and the libc++ that is actually linked. I have no visible object files linking to my executables as a result of compiling the BMIs anywhere...

could anyone shed some light on this?

germandiagogomez committed:
Add module partitions and implementation modules.

This commit adds support for implementation partitions
and modifies file conventions for a modularized target.

The conventions are:

  - your primary interface unit for your module should always
    be called 'module.cppm'. It should export module <target-name>.

  - your interface partitions can have any name supported by a module.
    For example: MyThings.cppm. The interface partition
    should declare 'export module <target-name>:MyThings'.

  - Your *importable* interface implementation units should end with
    'Impl.cppm'. For example 'MyThingsImpl.cppm'. This should
    'module <target-name>:MyThings' (without export).

  - Your *non-importable* implementation can have any name *with an
    extension .cpp*, not a .cppm, since implementation units are
    not importable. It is highly recommended, though, that if you
    have a single implementation file, it is called moduleImpl.cpp.

  You can see the project attached in Github draft PR:
  https://github.com/mesonbuild/meson/pull/14989 as an example that uses
  interface partitions, primary interface units, an implementation
  module (but not an implementation module partition, though it should
  work), multiple targets and import std.

  It can run a couple of tests when fully compiled and it resolves
  dependency order via scanner with Clang.

germandiagogomez commented:
> Oops sorry, I just saw you mention it in your opening post. Still, do you think it's actually worth it to merge any one-phase code? I think the changes required to do two-phase are minimal with your approach. When I tried to implement this a while ago, my problem with two-phase was that output file names were in no way predictible. But since you defined a naming scheme, this shouldn't be an issue. Let me know if you need any help.

Unofrtunately this weekend I had no time for this so I will continue the next one. 

Right now, when compiling, I generate the .pcm file in clang. It is one step as u said. This means I can do it by just adding ARGS to the existing compile command. I am not sure what should be done exactly to turn it into two-phase beyond the precompile flag but since it works, I was going to spend my time on gcc support and once I am done with it, revisit it.

> > Actually what I mean is: if I generate a bmi, should I generate also a library artifact and link against that libc++ instead of the original provided by the OS and -nostdlib++? Otherwise, my bmi should match the flags of the stock-provided OS or I am not sure how compatible it is. However, when I compiled the object file, it hardly included any code...

> In general, the object file built from a module is NOT in any way equal to the library itself. You can think of the pcm object file as all symbols that get defined, not just declared, when including the related header.

But it still generates object code. Module initialization code for example? I will need to investigate a bit what exactly goes into the object file and what it does not.

germandiagogomez commented:
@Jannik2099  I will not be able to work on this intensively in the next month or so.

Is there a possibility that maintainers could accept, as a first step, experimental support for 'import std' (only) and after that I will come back to the progress made by more core modules support as I have it (clang is working with the project attached).

The plan would be:

     1. import std added as experimental. (this PR)
     2. continue with clang modules (another PR) + optionally add gcc support.
     3. add gcc support separately in another PR based on the work on 2.
     
Please let me know what you think. Thanks!

jpakkane commented:
I looked into this [and found more bugs](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=122614). Let's see what upstream has to say on this.

mccakit commented:
We should only care about clang first, gcc should be a second priority

theoparis commented:
Is this still being worked on? I'm currently using CMake + Clang + Libc++ but I'd like to migrate to meson once support for this is added.

germandiagogomez commented:
> Is this still being worked on?

@theoparis 

I will give a second attempt to it during june, probably 2nd-3rd week. I want to have support for both clang and gcc. Realistically and with my amount of time, this is going to be an MVP for at least Clang, hope I can do something with gcc.

Not sure when or if it will ever enter mainstream though, that depends on core project people

dcbaker commented:
I would like to get C++ module support in, and I am a core developer.
